rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -0,0 +1,501 @@
+
+
+## Goals
+
+
+
+*   Parameter server training support for Keras compile/fit style training API
+*   Minimal code changes across usage with other strategies
+*   Minimal performance implications
+
+
+## Proposed options and solutions
+
+Let’s first take a look at the proposed user flow (on the coordinator). It is expected to be largely the same with other strategies (except for the strategy swap). Unless mentioned otherwise, the discussion here applies to the python program intended to be run on the coordinator.
+
+
+### User Journey
+
+With a dataset factory:
+
+
+```
+cluster_resolver = ...
+strategy = tf.distribute.experimental.ParameterServerStrategy(cluster_resolver)
+with strategy.scope():
+  preproc_stage = ... # Some Keras preproc layers
+  model = ... # Building a Keras model
+model.compile(optimizer=..., loss=...)  # `ClusterCoordinator` is created
+def dataset_fn(): 
+  ... # Make use of `preproc_stage` for transformation
+history = model.fit(dataset_fn, epochs=..., steps_per_epoch=...,  callbacks=[...])
+logging.info("result: %r", history)
+```
+
+
+with a dataset:
+
+
+```
+cluster_resolver = ...
+strategy = tf.distribute.experimental.ParameterServerStrategy(cluster_resolver)
+with strategy.scope():
+  preproc_stage = ... # Some Keras preproc layers
+  model = ... # Building a Keras model
+model.compile(optimizer=..., loss=...)  # `ClusterCoordinator` is created
+dataset = tf.data.Dataset.... # Make use of `preproc_stage` for transformation
+history = model.fit(dataset, epochs=..., steps_per_epoch=...,  callbacks=[...])
+logging.info("result: %r", history)
+```
+
+
+
+### Changes in `model.fit`
+
+This section discusses the changes needed to be made in `model.fit` API and assumes the reader has basic familiarity with Keras training APIs.
+
+
+#### `dataset` vs `dataset_fn` argument in `model.fit` API
+
+Current `model.fit` API takes a dataset from which an iterator is created, and the train function is built with this iterator. However, `ClusterCoordinator` only supports taking a no-argument function that returns a `Dataset`. We propose to have both options supported, with an initial focus on `dataset_fn` path due to its lighter implications on `model.fit` implementation.
+
+Previous discussion indicates that although an API modification is needed, the simplicity and less overhead may well justify such callable support, with a goal to support both.
+
+
+##### `dataset_fn` path
+
+In TF2 parameter server training, `ClusterCoordinator` naturally supports a dataset function to be passed in to `create_per_worker_dataset` API, which creates datasets on remote workers. By leveraging such data factory support, `model.fit` with `dataset_fn` can be implemented by subclassing the existing Keras `DataHandler` to provide a worker-distributed dataset for Keras to use (i.e. call `iter` on). Please see `DataHandler` section below for proposed changes.
```

Comment:
Worth noting for readers that DataHandler is an internal, private API.

Reply:
Added a note here.